Design/spectrum_number_and_workspace_index_abstraction.md

```diff
@@ -0,0 +1,212 @@
+- Algorithms that work with *detectors* use `DetectorIDsProperty` to translate from a user provided *detector ID* to an *index for a detector* (in an instrument).
+
+
+## Internals
+
+#### New functionality
+
+- Introduce a class that deals with translation and validation of indices.
+  This will replace code that is now in `MatrixWorkspace` (translation) and in many algorithms (validation).
+  Let us call this class `Translator` for the rest of this document.
+  - In contrast to the current situation this can be unit tested very well.
+  - Defines a consistent way of dealing with errors in the index specification.
+    This is basically also part of a consistent user interface.
+  - Reduces amount of code in algorithms and makes them more readable.
+- Provide a convenience function or method for declaring all relevant algorithm properties in a single statement.
+  - Reduces amount of code in algorithms and makes them more readable.
+  - Automatically gives consistent interface.
+- Spectrum numbers can be defined differently in each workspace.
+  As a consequence the object dealing with translation must be linked to a workspace.
+  For example, it could be a member of the workspace.
+- We should make sure that we do not negatively influence performance when a trivial set of indices is specified by the user.
+  Trivial can mean either the full range or a range in combination with a trivial mapping from spectrum number to workspace index (such as a simple offset of 1).
+
+To deal with MPI in a way that is hidden from both the user and most developers, `Translator` transforms a user input into, e.g., a `SpectrumIndexSet`.
+This is best illustrated by the example from above:
+
+```python
+# User input: SpectrumNumbers='2,3,7'
+indices = ws.makeSpectrumIndexSet(getProperty('SpectrumNumbers'))
+```
+
+Consider 3 cases:
+
+```python
+# Without MPI (or single MPI rank)
+Spectrum number        1 2 3 7 8
+Workspace index        0 1 2 3 4
+Global workspace index 0 1 2 3 4
+# indices contains 1,2,3
+
+# 2 MPI ranks (interleaved distribution of spectra)
+Spectrum number        1 2 3 7 8
+MPI rank               0 1 0 1 0
+Workspace index        0 0 1 1 2
+Global workspace index 0 1 2 3 4
+# rank 0: indices contains 1
+# rank 1: indices contains 0,1
+
+# 3 MPI ranks (block-wise distribution of spectra)
+Spectrum number        1 2 3 7 8
+MPI rank               0 0 1 1 2
+Workspace index        0 1 0 1 0
+Global workspace index 0 1 2 3 4
+# rank 0: indices contains 1
+# rank 1: indices contains 0,1
+# rank 2: indices is empty
+```
+
+Algorithms then use:
+
+```python
+for index in indices:
+  ws.mutableY(index) += offset
+```
+
+There are a few more complex cases.
+For example, in `LoadEventNexus` a vector of events labeled by detector ID needs to be inserted into a workspace.
+This requires *filtering* a vector with data based on a spectrum number of detector ID.
```

Comment:
Yes, that appears to be how it is done. It fails if there is not 1:1 mapping from detectors to spectra, but since this is in a load routine I think it is always true.